src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,342 @@
+    struct list_service_data *mdata = data;
+
+    mdata->node = node;
+
+    sol_ptr_vector_init(&mdata->service_list);
+
+    if (sol_netctl_add_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    return 0;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+
+    return -EINVAL;
+}
+
+static void
+close_list_services(struct sol_flow_node *node, void *data)
+{
+    static const char *errmsg = "Could not delete service monitor";
+    struct list_service_data *mdata = data;
+
+    if (sol_netctl_del_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    sol_ptr_vector_clear(&mdata->service_list);
+
+    return;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+}
+
+static int
+connection_handle(struct sol_flow_node *node, void *data, bool connect)
+{
+    struct connection_service_data *mdata = data;
+    const struct sol_ptr_vector *service_list;
+    struct sol_netctl_service *service;
+    uint16_t i = 0;
+    enum sol_netctl_service_state state;
+    int r;
+    const char *name;
+
+    static const char *state_msgs[] = {
+        [SOL_NETCTL_SERVICE_STATE_UNKNOWN] = "Unknown",
+        [SOL_NETCTL_SERVICE_STATE_IDLE] = "Idle",
+        [SOL_NETCTL_SERVICE_STATE_ASSOCIATION] = "Association",
+        [SOL_NETCTL_SERVICE_STATE_CONFIGURATION] = "Configuration",
+        [SOL_NETCTL_SERVICE_STATE_READY] = "Ready",
+        [SOL_NETCTL_SERVICE_STATE_ONLINE] = "Online",
+        [SOL_NETCTL_SERVICE_STATE_DISCONNECT] = "Disconnect",
+        [SOL_NETCTL_SERVICE_STATE_FAILURE] = "Failure",
+        [SOL_NETCTL_SERVICE_STATE_REMOVE] = "Remove",
+    };
+
+    service_list = sol_netctl_get_services();
+
+    while (service = sol_ptr_vector_get(service_list,i)) {
+        state = sol_netctl_service_get_state(service);
+        name = sol_netctl_service_get_name(service);
+
+        if (name && strcmp(name, mdata->service_name) == 0) {
+            if (state == SOL_NETCTL_SERVICE_STATE_IDLE && connect) {
+                sol_netctl_service_connect(service);
+            } else if ((state == SOL_NETCTL_SERVICE_STATE_READY) && (!connect)) {
+                  sol_netctl_service_disconnect(service);
+            }
+            state = sol_netctl_service_get_state(service);
+            r = sol_flow_send_string_packet(node,
+                    SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__OUT,
+                    state_msgs[state]);
+            if (r < 0)
+                SOL_WRN("Failed to send network service connect state");
+            return 0;
+        }
+        i++;
+    }
+
+    sol_flow_send_error_packet(node, ENOENT, "Did not found service name %s", mdata->service_name);
+
+    return 0;
+}
+
+static int
+connect_service_name(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    return connection_handle(node, data, true);
```

Comment:
with my proposal, the above `connection_handle()` would be converted into a `find_service_by_name()` and the logic is:

```c
struct sol_netctl_service *service = find_service_by_name(mdata->service_name);
if (!service) {
    sol_flow_send_error_packet(node, ENOENT, "Did not found service name %s", mdata->service_name);
    return 0; // there was no error in the callback itself, so no need to return error here
}

r = sol_netctl_service_connect(service);
if (r < 0)
    sol_flow_send_error_packet(node, -r, "Could not connect to service: %s: %s, mdata->service_name, sol_util_strerrora(-r));
return 0;
```